micall/drivers/sample.py

```diff
@@ -19,10 +19,31 @@
 from micall.core.denovo import denovo
 from micall.g2p.fastq_g2p import fastq_g2p, DEFAULT_MIN_COUNT, MIN_VALID, MIN_VALID_PERCENT
 from micall.utils.driver_utils import makedirs
+from contextlib import contextmanager
 
 logger = logging.getLogger(__name__)
 
 
+@contextmanager
+def open_files(**filemode):
+    files_opened = {}
+    for name, file_mode in filemode.items():
+        if file_mode is None:
+            files_opened[name] = None
+        else:
+            file = file_mode[0]
+            mode = file_mode[1]
+            files_opened[name] = open(file, mode)
+    try:
+        yield files_opened
+    finally:
+        for file in files_opened.values():
+            if file is None:
+                pass
+            else:
+                file.close()
```

Comment:
It is possible for `close()` to raise an error, when a buffered file fails to flush. It's probably easiest to test the scenario with mock objects.
Particularly tricky: you need to keep trying to close the remaining files, after an error. Then, raise an error. (Maybe the first one plus an extra message?)

Reply:
Ok, I think this case is covered (if I am catching the right exception, I think it should be either an OSError or an IOError), and tested as well.